guides/using-and-customizing-components.md

```diff
@@ -0,0 +1,205 @@
+
+**button.component.scss:**
+
+```sass
+.hc-checkbox {
+    input {
+        &:checked {}
+    }
+
+    &[disabled] input {}
+
+    .hc-label {}
+}
+```
+
+**button.scss:**
+
+```sass
+.hc-checkbox {
+    border: 1px solid slategray;
+
+    input {
+        background-color: white;
+
+        &:checked {
+            background-color: blue;
+        }
+    }
+
+    &[disabled] input {
+        background-color: lightgray;
+    }
+
+    .hc-label {
+        color: slategray;
+    }
+}
+```
+
+#### Step 4. Create mixins.
+
+Replace the CSS selectors in the new stylesheet with mixin declarations, flattening the structure as you go. Use BEM-like names for the mixins: `{component-name}__{subcomponent-name}--{component-state}`.
+
+**button.scss:**
+
+```sass
+@mixin checkbox() {
+    border: 1px solid slategray;
+}
+
+@mixin checkbox__input() {
+    background-color: white;
+}
+
+@mixin checkbox__input--checked() {
+    background-color: blue;
+}
+
+@mixin checkbox__input--disabled() {
+    background-color: lightgray;
+}
+
+@mixin checkbox__label() {
+    color: slategray;
+}
+```
+
+#### Step 5. Implement mixins.
+
+Use the mixins in the component stylesheet. Flatten the structure where possible, keeping in mind that the stylesheet is not encapsulated.
+
+**button.component.scss:**
+
+```sass
+@import '../sass/button.scss';
+
+.hc-checkbox {
+    @include checkbox();
+
+    input {
+        @include checkbox__input();
+
+        &:checked { @include checkbox__input--checked(); }
+    }
+
+    &[disabled] input {
+        @include checkbox__input--disabled();
+    }
+}
+
+.hc-label { @include checkbox__label(); }
+```
+
+#### Step 6. Review your code.
+
+Build the documentation and ensure that the component looks the same as it did before. Make sure it meets the following expectations:
+
+*   The component stylesheet should not include any CSS declarations. It should only have @import statements, CSS selectors, and @include statements. This ensures that all the component's styles can be used via mixins--none of them are bound to the component.
+*   The mixin stylesheet should not include any top-level CSS selectors. Each top-level item should be a mixin.
+*   The mixin stylesheet should not include any nested CSS selectors that would make confusing assumptions about the user's HTML structure. That is, the `[disabled]` attribute selector and pseudo-selectors like `:hover` or `:after` are okay, but class or tag names like `.hc-checked` and `input` should be avoided.
+*   Both stylesheets should be as flat as possible. Nesting CSS selectors in SCSS files creates extremely specific selectors, which are hard to override. Cashmere should make it easy to override a style rule if needed. CSS classes can be added to component markup to avoid deep nesting.
+*   If nesting more than 2 levels deep is unavoidable, an overridable SCSS variable (with `!default`) can be added to the mixins file to allow the user to set the rule before importing the mixins.
```

Comment:
Hmmm - yeah, realizing the same.  I actually think there are a lot of css classes in the library that included specific nested selectors, so this might be a bit of a challenge to work around.  I agree with @benjanderson that I think this is a good convention, but I fear it might be a little easier said than done.  Interested to hear your thoughts when you start digging into it @isaaclyman

Reply:
I realize it will take some work to flatten things out, but I really believe it will enhance the usability of the library. As someone who lost far too much time battling Bootstrap selectors back in the day...

But yeah, I'll keep you all informed on how it goes.